### PR TITLE
handle browser refresh

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

When the browser is refreshed, ensure the newly connected app calls runMain. The same vmservice can be reused and does not need to be updated.

## tests

Covered by existing cases

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart [hunk 1]
```diff
@@ -98,13 +99,19 @@ class WebFs {
...
   /// Retrieve the [DebugConnection] for the current application.
   Future<DebugConnection> runAndDebug() async {
-    final AppConnection appConnection = await _dwds.connectedApps.first;
-    appConnection.runMain();
-    return _dwds.debugConnection(appConnection);
+    final Completer<DebugConnection> firstConnection = Completer<DebugConnection>();
+    _connectedApps =  _dwds.connectedApps.listen((AppConnection appConnection) async {
+      appConnection.runMain();
+      if (!firstConnection.isCompleted) {
+        firstConnection.complete(await _dwds.debugConnection(appConnection));
```
Comment (mdebbar) [inline packages/flutter_tools/lib/src/build_runner/web_fs.dart]: This could lead to a race condition. Try doing the `await` before the `if`.

**Review by mdebbar — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/build_runner/web_fs.dart [hunk 2]
```diff
@@ -98,13 +99,20 @@ class WebFs {
     await _dwds.stop();
     await _server.close(force: true);
     await _chrome.close();
+    await _connectedApps?.cancel();
   }
 
   /// Retrieve the [DebugConnection] for the current application.
   Future<DebugConnection> runAndDebug() async {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_runner/web_fs.dart]: nit: async not needed.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/web_fs.dart]: I'll remove in https://github.com/flutter/flutter/pull/39951

**Pull request merged**